### PR TITLE
Allow UncaughtExceptionHandler to be overridden in client

### LINKED ISSUE #2331: OutOfMemoryError During BinMutations cause stuck BatchWriter
**Problem**
The first step in flushing mutations from the Batch Writer to the Tablet Servers involves binning mutations by destination tablet server. If the tablet server throws an OutOfMemoryError or any non-Exception based Throwable, the error will be thrown on the client side. The error will not be caught and will cause the bin thread to die without reporting the fact that an error occurred. This is unrecoverable and leaves the batch writer stuck waiting for mutations get flushed that never will.

**Affected versions**
1.10.x, main

**To Reproduce**
Not easily reproducible.
One could write a rogue iterator that throws an OutOfMemoryError when the metadata table is being scanned.

**Bug Location**
https://github.com/apache/accumulo/blob/5f20e38628abd6385a5ee652634f4d748ecd013f/core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java#L706

this:
`catch (Exception e) {`
should be changed to:
`catch (Throwable e) {`

### PR DESCRIPTION
**Pull request by dlmarion**
When an ExecutorService's task or Thread in Accumulo encounter an
unhandled exception the default UncaughtExceptionHandler will be
invoked. An Exception is logged, while an Error is logged and then
the VM is terminated. This change allows clients to override the
default uncaught exception handling logic so that the client VM is
not terminated when an Error happens in a background thread of a
client object.

Closes #2331 and supersedes #2346

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:176]:
```suggestion
        var clazz = Class.forName(uehClassName).asSubclass(UncaughtExceptionHandler.class);
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:181]: Could choose a more specific RTE. IAE seems appropriate, since it is instantiated by a value provided by the user. It's a bit late to be throwing that, since it's not directly in the builder, but it's probably still better than ISE. Could throw this earlier, in the builder.

File: core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java:28]: We should be taking care to avoid using JVM-wide static state. What if the user creates two clients with different exception handlers in the same JVM?

**Review by ctubbsii — CHANGES_REQUESTED** [review]
I like the approach overall, but I think some tweaks are need to ensure it only applies to that client only, and not all clients in the JVM. Also, I mention a few other possible tweaks below.

File: core/src/main/java/org/apache/accumulo/core/client/AccumuloClient.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/AccumuloClient.java:358]: With the code the way that it is, would it allow a user to put the class name in the client properties file? If so, then doing it this way would preclude that.

File: core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java:28]: Looking into this further, I think I want to just support a property in the client properties file and not on the AccumuloClient object. Everything in ThreadPools and Threads is static and I would need to change everything to be not-static and supply either a ClientContext or ServerContext object (which means I would have to change a ton more code). If we want to support a per-client UEH, then it might be easier to close this and update #2346 .

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java:28]: I think that other PR was trying to do a lot more than what I'm suggesting here, which is just the minimal amount of work to make Threads non-static and dangling off of ClientContext, so it can be have a per-client UEH. Yes, it would be larger than this PR. Maybe it could be done as a prerequisite, or even as a follow-on, to this one. But, I think the limitation of having one client control the UEH for all clients is substantial. Other than that limitation, I really like the direction this change took.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in 16bc28c

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in 16bc28c

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in 16bc28c

File: core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/threads/Threads.java]: I was able to pull some of the ideas from #2346 into this PR. 16bc28c should address the issue that you raised here.

File: core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:72]: Formatting is probably bad, but should be possible to use a lambda here.

```suggestion
    ScheduledFuture<?> future = ThreadPools.getClientThreadPools((t, e) ->
        LOG.error("context classloader cleanup thread has failed.", e)).createGeneralScheduledExecutorService(conf)
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java:379]: This cleanupThreadPool would probably be better left in ClientContext, and shut down when that is closed, rather than store it as an instance member here, where it's only ever used in the close method of this instance. It's very weird to *start* using resources for an object in that object's close method, which should be cleaning up resources. Putting it in ClientContext makes this effectively a singleton again (per Client, not per JVM, though, but that's still better than per-ConditionalWriter) and its own cleanup will be a bit more sane there, I think.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ScannerIterator.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ScannerIterator.java:70]: Same as my previous comment, this could be moved to ClientContext and cleaned up on `client.close()`. No need for the Cleaner.

File: core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: I don't know if it's easy to do in all cases, but I was thinking that instead of this, it can share the same sort of code as for the client-side cases.

Essentially, ClientContext would have a `.threadPools()` method that returns an instance of `ThreadPools` constructed using the UEH provided when AccumuloClient was built (default if not provided). When ServerContext is constructed, it just passes the built-in UEH for the server-side code to ClientContext's constructor in `super()`.

So, instead of calling `ThreadPools.getServerThreadPools().whatever`, you'd just use `context.threadPools().whatever` everywhere, regardless of client or server context. That way, there's no static instance of ThreadPools anywhere, but everything still works as expected. (`context.threadPools()` is also shorter than `ThreadPools.getServerThreadPools()`)

**Review by ctubbsii — COMMENTED** [review]
I like these changes, but I think there's room for further improvement. Some could be done later as well.

File: core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: I don't know that in all cases server-side that the ServerContext is available to the code. I would have to wire it up in all cases where it's not reachable.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: To be clear, only ClientContext would be necessary to wire up. ServerContext isn't available in the core jar. However, it is a subclass of ClientContext, so it would still work if only a ClientContext. But even then, ClientContext might not be available everywhere, and might need to be wired up... that's one of the things we've been fighting against to avoid static state in the JVM and the problems that come with it, so it's not a new issue. But, I was just hoping we could try to avoid a few additional occurrences of it that we'd have to fix later.

In any case, that could be treated as a follow-on. I think the main things to address before this is merged in is to move the static thread pools that were converted to instance members for cleanup into ClientContext, so they exist in a Closeable object and don't need a separate Cleaner, and so they can be reused for the cleanup tasks they perform across multiple instances of the objects they clean up.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: Won't they still need a Cleaner in the case where the client is not closed?

File: core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java]: Addressed in ec0e7cf

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java]: Addressed in ec0e7cf

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ScannerIterator.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ScannerIterator.java]: Addressed in ec0e7cf

File: core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: I thought the client context already has a cleaner that ensures it is closed, because of the legacy Instance/Connector stuff that may not clean up the client context.

However, even if it doesn't, Cleaner's (replacing finalizers) are a hack to close things during garbage collection, and are primarily useful for objects that don't explicitly get closed. For objects whose lifecycle is user-facing, like our Closeable clients, we don't need to specify a Cleaner... we (or the user) can just close them. Java helps us with this by giving us warnings about resource leaks and potential resource leaks for unclosed Closeable objects.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:801]: I think these cleaners can go away, based on my previous comments.

**Review by ctubbsii — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: I see the something in SingletonReservation that closes AccumuloClients if that is what you are referencing. I will remove the Cleaners from ClientContext.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in 180654d

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:787]: These should be ordered before the `singletonReservation.close()` call above them. Also, I noticed the other thread pools held by this class are lazily instantiated, so could be null. Should these be lazily instantiated as well? It would save some resources when using a client for things that doesn't require them.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:178]: The way this is being read from the config is a bit weird. It's reading a ClientProperty from an AccumuloConfiguration, which means that the client property would have had to exist in the `accumulo.properties` file, but not documented as a server-side property. The property is explicitly named `client.uncaught.exception.handler`, but the way it is loaded, it will set it for the server-side as well. I think this would be cleaner if it were renamed to `instance.uncaught.exception.handler` and put in `Property.java` with other server properties. You could even use ClassLoaderUtil directly, or via ConfigurationTypeHelper to load the class from the configuration.

If you wanted a separate client-side property name, you could do that (see how ClientConfConverter translates ZooKeeper host names, for an example), but I don't think that'd be necessary.

File: core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java:73]: It is. Thanks.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:787]: I can move the `close()` calls up. Regarding lazy instantiation, what other thread pools were you referencing? The class javadoc says `Any state in this object should be available at the time of its construction.`, does that on mean configuration?

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:178]: I think I'm misunderstanding something as I'm not familiar with the new Accumulo client in 2.x. Does the Accumulo client in 2.x use the `accumulo-client.properties` file? I was thinking that it did, now I'm thinking that it does not. Is that file only used for the shell and utilities?

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:787]: I was referring to `thriftTransportPool`, a few lines above these. This class has evolved since I originally wrote that javadoc. While that was the original intent, I'm not sure it's entirely necessary or practical given how much state we're using in here. What is most important is that everything is available in the sense that no additional API calls are necessary to further initialize the state, that all of its functionality is ready after constructing. Lazy initialization would still be compatible with that looser definition of availability.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:178]: The client does use that file, but the server-side doesn't. This code is shared by both client and server code because `ClientContext` is the `Impl` class for `AccumuloClient`, and it is the class that is returned from the `AccumuloClient` builder. `ClientContext` is also the super-class of `ServerContext`, because servers can also perform client operations.

The builder for `AccumuloClient` can initialize the configuration used in `ClientContext` from the client configuration file. However, the `ServerContext` initializes the configuration used in `ClientContext` from the server's `conf/accumulo.properties` file (which clients cannot be expected to have access to). Complicating the situation, we have two different versions of client properties, and have a translator tool to convert one into the other. And, to share code between the two use cases, we present the client configuration file to `ClientContext` in an `AccumuloConfiguration` view.

Because this class is the super-class of `ServerContext`, we would need it to be able to read the value from the `conf/accumulo.properties` file for the server-side. Users can have the option of putting the same configuration property in their own `accumulo-client.properties` file, if they wish (and that's the main use case of this PR). For other properties where that happens, we try to keep the same name, so it's easier. However, for historical reasons, some of the properties, like the one for setting the ZooKeeper hosts, have a different property name that we translate if we see it.

Yeah, it's complicated... but that's the trade-off for preserving deprecated functionality, properties, key names, etc. I would *love* to strip out the old stuff, and dramatically simplify a lot of this, but that has been resisted in the past (understandably). I'm hoping we can clean it up with a non-LTM 3.0 release before we start working on an LTM 3.x version.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:178]: @ctubbsii - Thanks for the explanation, that helped. I'm about to push a commit that:

1. addresses your comments above about lazy initialization of the thread pools
2. removes the UEH handler property entirely
3. Sets the UEH to `Threads.UEH` when `ClientContext` is initialized from `ServerContext`
4. Allows the client to set a UEH class on the AccumuloClient when building it
5. Uses a default UEH in the client if one is not supplied that logs the error. This should provide some backwards compatibility for clients so the client VM does not get terminated.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:205]: Could consider creating this in the constructor, then would not need to synchronize the method.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:124]: Why was volatile removed?  W/o volatile or synchronization, changes may not be seen.  Like `ensureOpen()` may not see changes.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:794]: Either need to make this variable volatile or synchronize this method to ensure changes to the variable made by different threads are seen.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:205]: @keith-turner I believe that's what he did in a previous iteration, but I suggested he lazily instantiate, to follow the convention established with the thriftTransportPool which avoids allocating resources which might not be used by the context.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:794]: @keith-turner  Do you believe this is an existing problem with the thriftTransportPool also?

File: server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java
Comment (dlmarion) [reply server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java:130]: That would work in the cases where ServerContext is reachable. They are functionally equivalent.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:205]: That's correct.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:124]: I think it was lost in a prior change, which I then reverted. I will add it back.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:794]: I synchronized the `close` method in 47e90dca559f02683c7c9f58ae366a422367a2d7

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in 47e90dca559f02683c7c9f58ae366a422367a2d7

File: server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java
Comment (ctubbsii) [reply server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java:130]: > That would work in the cases where ServerContext is reachable. They are functionally equivalent.

While these are functionally equivalent now, it would save effort later, trying to wire in context where needed to get rid of the static references to these resources, in case we wanted to make the server-side UEH configurable at some point.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:210]:
```suggestion
  public synchronized void executeCleanupTask(Runnable r) {
    ensureOpen();
    if (cleanupThreadPool == null) {
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:202]:
```suggestion
  public synchronized Future<List<KeyValue>>
      submitScannerReadAheadTask(Callable<List<KeyValue>> c) {
    ensureOpen();
    if (scannerReadaheadPool == null) {
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:794]: I'm not sure it's necessary, since the `closed` variable is volatile again. As long as you add `ensureOpen();` to the two methods that retrieve the pools. That said, it's not like we're calling `close()` so frequently that synchronizing it is going to matter.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:244]: Is this getter method ever used outside the constructor? If not, it can be removed, along with the field it gets.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Method removed in a6236ae

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Addressed in a6236ae

**Review by ctubbsii — APPROVED** [review]
I think this is good, but since @keith-turner was also reviewing it, I'd wait on his response to the recent updates/comments before merging.

File: server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java
Comment (keith-turner) [reply server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionCoordinator.java:130]: If we can use the context method instead of the static method that would be nice, but that can be follow on work.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:203]: As long as closed() is synchronized, that addresses the main concern I had.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:789]: Yeah closed does not need to be efficient, no need to worry about thread contention here.

**Review by keith-turner — APPROVED** [review]

**Pull request merged**
